### PR TITLE
LPS 101640 Modernize addParams JS utility

### PR DESCRIPTION
**Pull request by mateomustapic**

### CODE REVIEW

**Comment by mateomustapic** [comment]
ci:test:sf

**Comment by mateomustapic** [comment]
ci:test:stable

**Comment by mateomustapic** [comment]
ci:test:relevant

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/add_params.es.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+	it('throws error if url parameter is not a string', () => {
+		expect(() => addParams('foo', 0)).toThrow(
+			'Parameter url must be a string'
+		);
+	});
+
+	it('Adds the parameters to the portlet URL', () => {
+		expect(addParams({bar: 'foo', foo: 'bar'}, sampleUrl)).toEqual(
+			'http://sampleurl.com?foo=bar&bar=foo'
```
Comment (julien) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/add_params.es.js:38]: @mateomustapic This test is failing, and my guess is because you use `{bar: 'foo', foo: 'bar'}` as the first argument to the `addParams` call so I would expect the return value to be `'http://sampleurl.com?bar=foo&foo=bar'`.

This will probably happen too with the two last tests in this file.

Comment (mateomustapic) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/add_params.es.js:38]: fixed, resending now

**Comment by mateomustapic** [comment]
ci:stop

**Comment by mateomustapic** [comment]
ci:test:sf

**Comment by mateomustapic** [comment]
ci:test:stable

**Comment by mateomustapic** [comment]
ci:test:relevant

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
+/**
```
Comment (julien) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js:1]: @mateomustapic I still have one small doubt.

We're using `es.js` for the file extension, and (I think) we don't do that anymore now.
But I also understand that you chose this extension because some files in this (OSGI) module still use `es.js`.

Since these are new files that we're creating I'm not 100% sure about what to do?

We should probably talk with @wincent's to get his point of view.

Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js:1]: Ideally, we want things to be named `.js` from now on, although there's no great hurry. Personally, I would start naming all new files `.js`, even in modules that still have a lot of `.es.js` in them. Eventually, we will move them all.

But there are some gotchas:

- In old modules like frontend-js-web which use AUI you can run into problems with the `.npmbundlerc` config and specifically the `ignore` property. See @markocikos's PR [here](https://github.com/brianchandotcom/liferay-portal/pull/87585) which just got merged last night.

Note that frontend-js-web used to *ignore* all `.js` files, meaning they wouldn't get processed by Babel and you couldn't safely use ES6 features in them. `.es.js` files were explicitly excluded from that via a whitelist approach. With Marko's PR, we got rid of the whitelist and are instead blacklisting the specific AUI-using and other legacy modules that should not go through Babel.

Long story short: you can safely use `.js` now in this module. If the inconsistency really bothers you, you could make an argument for leaving these two new files as `.es.js`. Personally, I'd use `.js`.

Comment (julien) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js:1]: Thanks @wincent!
In our case as you said, we can used `.js` without a problem so that's what we'll do.

**Comment by wincent** [comment]
@mateomustapic: As a matter of habit, can you please make sure *all* PRs you send have a description, even if its brief? [Details and rationale](https://github.com/liferay/liferay-frontend-guidelines/blob/master/general/pull_requests.md#write-useful-descriptions-and-titles).

In this case, for example, what does "modernize" mean? A one sentence description should suffice, saying this creates a vanilla-JS replacement for an old API (name/link the old API).

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/add_params.es.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		);
+	});
+
+	it('Sets anchor hash after base url and parameters', () => {
+		expect(
+			addParams({bar: 'foo', foo: 'bar'}, anchorHashSampleUrl)
+		).toEqual('http://sampleurl.com?bar=foo&foo=bar#foo');
+	});
+
+	it('Sets portlet url when base url has parameters', () => {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/add_params.es.js:54]: For consistency, none of these should start with capital letters. This might be something that we can add a lint for. Created a ticket for that: https://github.com/liferay/eslint-config-liferay/issues/172

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/add_params.es.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	});
+
+	it('throws error if url parameter is not a string', () => {
+		expect(() => addParams('foo', 0)).toThrow(
+			'Parameter url must be a string'
+		);
+	});
+
+	it('Adds the parameters to the portlet URL', () => {
+		expect(addParams({bar: 'foo', foo: 'bar'}, sampleUrl)).toEqual(
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/add_params.es.js:37]: Just a little nit, but the mixing of foo/bar here makes this harder to read than it needs to be ie. "bar is foo and foo is bar" is at best distracting and at worst confusing.

Can we pick something more "real", like:

- `{limit: 10, order: 'desc'}` 
- `{color: 'red', size: 'large'}`

Or whatever else strikes your fancy.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	if (typeof params !== 'object' && typeof params !== 'string') {
+		throw new TypeError('Parameter params must be an object or string');
+	}
+
+	if (typeof url !== 'string') {
+		throw new TypeError('Parameter url must be a string');
+	}
+
+	if (typeof params === 'object') {
+		const paramKeys = Object.keys(params);
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js:32]: This is invalid. Will throw if `params` is `null` (because `Object.keys(null)` throws, and `typeof null === 'object'`).

You pretty much never want to write `typeof x === 'object'` without also writing `&& x`.

I thought we had an issue to create a lint for this but I can't find it, so I made: https://github.com/liferay/eslint-config-liferay/issues/173

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+		params = paramKeys
+			.map(key => {
+				return `${encodeURIComponent(key)}=${encodeURIComponent(
+					params[key]
+				)}`;
+			})
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js:39]: Whenever you need both keys and values, you should probably use `Object.entries` instead of `Object.keys`. Additionally that would enable you to get rid of the ugly `paramKeys` temporary variable ie.

```js
params = Object.entries(params).map(([key, value]) => (
    `${encodeURIComponent(key)}=${encodeURIComponent(value)}`
));

Might even be narrow enough to avoid the excessive wrapping of the template literal.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+			.join('&');
+	}
+	else {
+		params = params.trim();
+	}
+
+	let loc = url || location.href;
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js:46]: So we insist that `url` be a string earlier on, but if it is empty, we fall back to `location.href`. Do people often call this with an empty string? It might be nicer to allow them to just omit the paramter.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+	let finalUrl = loc;
+
+	if (params) {
+		let anchorHash;
+
+		if (loc.indexOf('#') > -1) {
+			const locationPieces = loc.split('#');
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js:54]: The variable naming is a bit of a mess here. On the one hand we use an abbreviation like `loc` (please don't do that), then we use `locationPieces` which seems oddly specific.

Just say you rename `loc` to `location`, you can avoid this awkward temporary variable entirely with:

```
([location, fragment] = location.split('#'));
```

(Note: I also renamed `anchorHash` to `fragment` because that's what it is — [see here](https://en.wikipedia.org/wiki/URL#Syntax)).

Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js:54]: Another thing to note here; if we ever got a malformed URL with more than one `#` in it, this `split()` would truncate the everything from the second `#` onwards. Do we care about that? Might want to make it handle that edge case more gracefully.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+		if (loc.indexOf(params) === -1) {
+			finalUrl = loc + params;
+
+			if (anchorHash) {
+				finalUrl += `#${anchorHash}`;
+			}
+			if (!url) {
+				location.href = finalUrl;
+			}
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js:75]: This seems bad. If we assign to `location.href`, the browser will navigate away. I expect a function like this that *returns a URL* to be free of side-effects...

**Comment by julien** [comment]
@mateomustapic I tried to analyse the output of the test failures, which don't seem related, but just in case and once we've applied the requested changes, we should ask QA for some help on this.

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/add_params.es.js (hunk 4 above)
Comment (mateomustapic) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/add_params.es.js:37]: I agree on that, I will modify these parameters and make them more readable.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js (hunk 8 above)
Comment (mateomustapic) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js:54]: This chunk of code I rewrote as it was in frontend-aui-web. 
I will improve this code and also work on variable naming here.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js (hunk 6 above)
Comment (mateomustapic) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.es.js:39]: This is definitely useful, thank you for this suggestion.

**Comment by mateomustapic** [comment]
Hey @wincent  looking at this code, i think we could be able to use the URL api It does the same as this utility. So instead of modifying the code by suggestions in this PR we could use URL api, what do you think?

**Comment by wincent** [comment]
> Hey @wincent looking at this code, i think we could be able to use the URL api It does the same as this utility. So instead of modifying the code by suggestions in this PR we could use URL api, what do you think?

Sounds like a splendid idea.

**Comment by julien** [comment]
closing this one since changes have been requested and are still in progress

**Comment by julien** [comment]
ci:close
